Prepare for TensorPipe separating its CUDA-specific headers

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59788 Prepare for TensorPipe separating its CUDA-specific headers**
* #59377 Move CUDA-related stuff of TP agent to separate file
* #59376 Make CUDA serde support for TP agent pluggable

This one line is all we need to "migrate" PyTorch to the "new API" of TensorPipe that splits the CUDA-specific stuff in a separate top-level header. (The motivation behind that is that it will allow us to "stack" the CUDA code on top of the CPU one).

Differential Revision: [D28875277](https://our.internmc.facebook.com/intern/diff/D28875277/)

lw committed:
Prepare for TensorPipe separating its CUDA-specific headers

This one line is all we need to "migrate" PyTorch to the "new API" of TensorPipe that splits the CUDA-specific stuff in a separate top-level header. (The motivation behind that is that it will allow us to "stack" the CUDA code on top of the CPU one).

Differential Revision: [D28875277](https://our.internmc.facebook.com/intern/diff/D28875277/)

[ghstack-poisoned]

lw committed:
Update on "Prepare for TensorPipe separating its CUDA-specific headers"


This one line is all we need to "migrate" PyTorch to the "new API" of TensorPipe that splits the CUDA-specific stuff in a separate top-level header. (The motivation behind that is that it will allow us to "stack" the CUDA code on top of the CPU one).

Differential Revision: [D28875277](https://our.internmc.facebook.com/intern/diff/D28875277/)

[ghstack-poisoned]

lw committed:
Update on "Prepare for TensorPipe separating its CUDA-specific headers"


This one line is all we need to "migrate" PyTorch to the "new API" of TensorPipe that splits the CUDA-specific stuff in a separate top-level header. (The motivation behind that is that it will allow us to "stack" the CUDA code on top of the CPU one).

Differential Revision: [D28875277](https://our.internmc.facebook.com/intern/diff/D28875277/)

[ghstack-poisoned]

lw committed:
Update on "Prepare for TensorPipe separating its CUDA-specific headers"


This one line is all we need to "migrate" PyTorch to the "new API" of TensorPipe that splits the CUDA-specific stuff in a separate top-level header. (The motivation behind that is that it will allow us to "stack" the CUDA code on top of the CPU one).

Differential Revision: [D28875277](https://our.internmc.facebook.com/intern/diff/D28875277/)

[ghstack-poisoned]

mrshenli reviewed (commented):
LGTM. 

The following error does not look relevant, but it doesn't seem to occur on master. 

```
Jun 11 14:49:14   test_forward_sync_script (__main__.TensorPipeRemoteModuleTestWithSpawn) ... Caught exception: 
Jun 11 14:49:14 Traceback (most recent call last):
Jun 11 14:49:14   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 516, in run_test
Jun 11 14:49:14     getattr(self, test_name)()
Jun 11 14:49:14   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 400, in wrapper
Jun 11 14:49:14     fn()
Jun 11 14:49:14   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/dist_utils.py", line 100, in new_test_method
Jun 11 14:49:14     return_value = old_test_method(self, *arg, **kwargs)
Jun 11 14:49:14   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/distributed/nn/api/remote_module_test.py", line 203, in test_forward_sync_script
Jun 11 14:49:14     dst_worker_name, modes=[ModuleCreationMode.MODULE_CTOR_WITH_INTERFACE]
Jun 11 14:49:14   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/distributed/nn/api/remote_module_test.py", line 123, in _create_remote_module_iter
Jun 11 14:49:14     _module_interface_cls=MyModuleInterface,
Jun 11 14:49:14   File "/opt/conda/lib/python3.6/site-packages/torch/distributed/nn/api/remote_module.py", line 241, in __init__
Jun 11 14:49:14     fut.wait()  # Ensure remote_module_cls is available on remote side.
Jun 11 14:49:14   File "/opt/conda/lib/python3.6/site-packages/torch/distributed/rpc/internal.py", line 218, in _handle_exception
Jun 11 14:49:14     raise result.exception_type(result.msg.encode("utf-8").decode("unicode_escape"))
Jun 11 14:49:14 OSError: On WorkerInfo(id=1, name=worker1):
Jun 11 14:49:14 OSError("Can't get source for <function _remote_forward at 0x7f07998932f0>. TorchScript requires source access in order to carry out compilation, make sure original .py files are available.",)
Jun 11 14:49:14 Traceback (most recent call last):
Jun 11 14:49:14   File "/opt/conda/lib/python3.6/site-packages/torch/_utils_internal.py", line 59, in get_source_lines_and_file
Jun 11 14:49:14     sourcelines, file_lineno = inspect.getsourcelines(obj)
Jun 11 14:49:14   File "/opt/conda/lib/python3.6/inspect.py", line 955, in getsourcelines
Jun 11 14:49:14     lines, lnum = findsource(object)
Jun 11 14:49:14   File "/opt/conda/lib/python3.6/inspect.py", line 786, in findsource
Jun 11 14:49:14     raise OSError('could not get source code')
Jun 11 14:49:14 OSError: could not get source code
```

lw committed:
Update on "Prepare for TensorPipe separating its CUDA-specific headers"


This one line is all we need to "migrate" PyTorch to the "new API" of TensorPipe that splits the CUDA-specific stuff in a separate top-level header. (The motivation behind that is that it will allow us to "stack" the CUDA code on top of the CPU one).

Differential Revision: [D28875277](https://our.internmc.facebook.com/intern/diff/D28875277/)

[ghstack-poisoned]

lw commented:
> The following error does not look relevant, but it doesn't seem to occur on master.

I'm very tempted to dismiss it as a flake because that's a Python error whereas here I just add an empty C++ header. Moreover, I add this header in a CUDA-only file, whereas that error seems to appear in CPU code. Anyways, I reexported my PRs, let's see if it occurs again.
